torch/lib/THD/base/data_channels/DataChannelNccl.cpp

```diff
@@ -0,0 +1,740 @@
+  THCudaCheck(cudaSetDevice(curDevice));
+
+  // Move into the hash table
+  _ncclCommsAndEvents[groupId] = std::move(std::make_pair(std::move(comms),
+                                                          std::move(events)));
+
+  return std::make_pair(_ncclCommsAndEvents[groupId].first.get(),
+                        _ncclCommsAndEvents[groupId].second.get());
+}
+
+
+// Helper function that checks the input and output tensors for validity
+void DataChannelNccl::_tensorCheckHelper(
+    const std::vector<at::Tensor>& input,
+    const std::vector<at::Tensor>& output,
+    size_t outputOverInput) {
+
+  if (input.size() <= 0) {
+    throw std::runtime_error("Input tensor sequence cannot be empty");
+  }
+  if (input.size() != output.size()) {
+    throw std::runtime_error("Input tensor sequence should have the same size "
+                             "as output tensor sequence");
+  }
+  if (input.size() > _numGPUs) {
+    throw std::runtime_error("The number of input tensors is larger than "
+                             "the number of available GPUs");
+  }
+
+  // To make sure each tensor is on separate devices
+  std::unordered_set<int> usedDevices;
+  usedDevices.reserve(input.size());
+
+  uint64_t inputNumElement = input[0].numel();
+  auto elementType = input[0].type().scalarType();
+
+  for (size_t i = 0; i < input.size(); ++i) {
+
+    //  Check to make sure it's a GPU dense tensor
+    if (!(input[i].type().isCuda() && !input[i].type().isSparse() &&
+          output[i].type().isCuda()  && !output[i].type().isSparse())) {
+      throw std::runtime_error("Only CUDA dense tensor is supported for NCCL "
+                               "collective operations");
+    }
+    // Check the tensor type is identical
+    if (input[i].type().scalarType() != elementType ||
+        output[i].type().scalarType() != elementType) {
+      throw std::runtime_error("Expecting all GPU tensors to have identical "
+                               "type");
+    }
+    // Check the input tensor size is identical
+    if (input[i].numel() != inputNumElement) {
+      throw std::runtime_error("Expecting all input tensors to have identical "
+                               "number of elements");
+    }
+    // Check the output tensor size equals to input tensor size
+    if (output[i].numel() != inputNumElement * outputOverInput) {
+      throw std::runtime_error("The number of elements of output tensor does "
+                               "not match the number of elements of the input "
+                               "tensor");
+    }
+    // Contiguous verification
+    if (!input[i].is_contiguous() || !output[i].is_contiguous()) {
+      throw std::runtime_error("Expecting all GPU tensors to be contiguous");
+    }
+    // Device verification
+    if (usedDevices.find(input[i].get_device()) != usedDevices.end()) {
+      throw std::runtime_error("Expecting inputs on different GPU devices");
+    }
+
+    usedDevices.insert(input[i].get_device());
```

Comment:
insert is making it's own check of whether inserted element is already present in the set, so to avoid doing it twice you can check .insert() return https://stackoverflow.com/a/41152203